CMakeLists.txt

```
@@ -0,0 +1,136 @@
+cmake_minimum_required(VERSION 3.0)
+
+SET(CMAKE_SYSTEM_NAME Generic)
+SET(CMAKE_C_COMPILER avr-gcc)
+SET(CMAKE_CXX_COMPILER avr-g++)
+SET(CMAKE_CAT cat)
+
+SET(CSTANDARD "-std=gnu11")
+SET(CPPSTANDARD "-std=gnu++11")
+
+SET(WARN "-Wall -Wextra")
+SET(CPPTUNING "-fno-exceptions -fno-threadsafe-statics")
+SET(OPT "-Os -ffunction-sections -fdata-sections -flto -fno-fat-lto-objects")
+SET(MCU "-mmcu=atmega32u4")
+
+add_compile_definitions(
+	F_CPU=16000000
+	ARDUINO=10805
+	ARDUINO_AVR_prusa_mm_control
+	ARDUINO_ARCH_AVR
+	USB_VID=0x2c99
+	USB_PID=0x0004
+	USB_MANUFACTURER="Unknown"
+	USB_PRODUCT="Original Prusa i3 MK3 Multi Material 2.0 upgrade"
+)
+
+SET(CFLAGS "${OPT} ${WARN} ${CSTANDARD} ${MCU}")
+SET(CXXFLAGS "${OPT} ${WARN} ${CPPSTANDARD} ${CPPTUNING} ${MCU}")
+SET(ASMFLAGS "${OPT} ${WARN} ${MCU}")
+
+SET(CMAKE_C_FLAGS ${CFLAGS})
+SET(CMAKE_CXX_FLAGS ${CXXFLAGS})
+SET(CMAKE_ASM_FLAGS ${ASMFLAGS})
+SET(CMAKE_EXE_LINKER_FLAGS "-fuse-linker-plugin -Wl,--gc-sections")
+
+
+project(MM-control-01 C CXX ASM)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+include(GetGitRevisionDescription)
+
+get_git_head_revision(GIT_HEAD_REF GIT_COMMIT_HASH)
+git_commits_in_master(GIT_COMMITS_MASTER)
+git_local_changes(GIT_LOCAL_CHANGES)
+
+configure_file(
+${CMAKE_SOURCE_DIR}/MM-control-01/version.h.in
+version.h
+)
+# The 'real' git information file
+set(GITREV_BARE_FILE dirty.h)
+# The temporary git information file
+set(GITREV_BARE_TMP dirty-tmp.h)
+set(GITREV_FILE ${CMAKE_BINARY_DIR}/${GITREV_BARE_FILE})
+set(GITREV_TMP ${CMAKE_BINARY_DIR}/${GITREV_BARE_TMP})
+
+set(SOURCES 
+	MM-control-01/abtn3.c
+	MM-control-01/adc.c
+	MM-control-01/motion.cpp
+	MM-control-01/main.cpp
+	MM-control-01/tmc2130.c
+	MM-control-01/permanent_storage.cpp
+	MM-control-01/Buttons.cpp
+	MM-control-01/uart.cpp
+	MM-control-01/shr16.c
+	MM-control-01/mmctl.cpp
+	MM-control-01/spi.c
+	core/abi.cpp
```

Maybe isolate the core into a stand-alone library?
```suggestion
add_library(core STATIC
```

---

Using static libraries increase build time. This is no problem in this project, as I have compile time around 1s. But libraries overhead in bigger projects can be noticeable - e.g. 100MB static library needs to be linked and written to HDD.

---

@mkbel So you'd rather trade clarity for virtually unnoticable build-time reduction? 

https://sourceware.org/binutils/docs/binutils/ar.html

> GNU ar can optionally create a thin archive, which contains a symbol index and references to the original copies of the member files of the archive. This is useful for building libraries for use within a local build tree, where the relocatable objects are expected to remain available, and copying the contents of each object would only waste time and space.